### PR TITLE
Support icon color for text entry and column

### PR DESCRIPTION
**Pull request by aymanalareqi**
- [x] Changes have been thoroughly tested to not break existing functionality.
- [x] New functionality has been documented or existing documentation has been updated to reflect changes.
- [x] Visual changes are explained in the PR description using a screenshot/recording of before and after.

This PR add ability to customize icon color for TextColumn component icon like the below screenshot
Screenshot 2023-11-08 at 8 09 31 AM

also add ability to customize icon color for infolist TextEntry component as below screenshot
Screenshot 2023-11-08 at 8 43 51 PM

### CODE REVIEW

File: packages/support/src/Contracts/HasIconColor.php
Comment (danharrin) [inline packages/support/src/Contracts/HasIconColor.php:1]: Please remove this, it's not being used

Comment (zepfietje) [reply packages/support/src/Contracts/HasIconColor.php:1]: Mark as ready for review when done, @aymanalareqi.

**Comment by aymanalareqi** [comment]
@danharrin  @zepfietje  I have removing packages/support/src/Contracts/HasIconColor.php but there is some failing in phpstan

I think the failing phpstan not from this PR

**Comment by wychoong** [comment]
can you add support for Infolist entries as well please
https://filamentphp.com/docs/3.x/infolists/entries/text#adding-an-icon

**Comment by aymanalareqi** [comment]
> can you add support for Infolist entries as well please https://filamentphp.com/docs/3.x/infolists/entries/text#adding-an-icon

added

**Comment by aymanalareqi** [comment]
@danharrin  @zepfietje  I have added screenshot url in docs ,Should I added the image or they will be added automatically

File: packages/tables/resources/views/columns/text-column.blade.php [hunk 1]
```diff
@@ -77,23 +77,24 @@
                         $copyMessageDuration = $getCopyMessageDuration($state);
                         $fontFamily = $getFontFamily($state);
                         $icon = $getIcon($state);
+                        $iconColor = $getIconColor($state);
```
Comment (zepfietje) [inline packages/tables/resources/views/columns/text-column.blade.php]:
```suggestion
                        $iconColor = $getIconColor($state) ?? $color;
```

File: packages/infolists/resources/views/components/text-entry.blade.php [hunk 2]
```diff
@@ -87,17 +88,17 @@
 
                                 $iconClasses = \Illuminate\Support\Arr::toCssClasses([
                                     'fi-in-text-item-icon h-5 w-5 shrink-0',
-                                    match ($color) {
+                                    match ($iconColor ?? $color) {
```
Comment (zepfietje) [inline packages/infolists/resources/views/components/text-entry.blade.php]:
```suggestion
                                    match ($iconColor) {
```

File: packages/infolists/resources/views/components/text-entry.blade.php [hunk 3]
```diff
@@ -87,17 +88,17 @@
...
                                         'gray', null => 'text-gray-400 dark:text-gray-500',
                                         default => 'text-custom-500',
                                     },
                                 ]);
 
                                 $iconStyles = \Illuminate\Support\Arr::toCssStyles([
                                     \Filament\Support\get_color_css_variables(
-                                        $color,
+                                        $iconColor ?? $color,
```
Comment (zepfietje) [inline packages/infolists/resources/views/components/text-entry.blade.php]:
```suggestion
                                        $iconColor,
```

File: packages/infolists/resources/views/components/text-entry.blade.php [hunk 4]
```diff
@@ -87,17 +88,17 @@
...
                                     },
                                 ]);
 
                                 $iconStyles = \Illuminate\Support\Arr::toCssStyles([
                                     \Filament\Support\get_color_css_variables(
-                                        $color,
+                                        $iconColor,
                                         shades: [500],
-                                    ) => $color !== 'gray',
+                                    ) => ($iconColor ?? $color) !== 'gray',
```
Comment (zepfietje) [inline packages/infolists/resources/views/components/text-entry.blade.php]:
```suggestion
                                    ) => $iconColor !== 'gray',
```

File: packages/tables/resources/views/columns/text-column.blade.php [hunk 5]
```diff
@@ -77,23 +77,24 @@
...
                         $icon = $getIcon($state);
+                        $iconColor = $getIconColor($state) ?? $color;
                         $itemIsCopyable = $isCopyable($state);
                         $size = $getSize($state);
                         $weight = $getWeight($state);
 
                         $iconClasses = \Illuminate\Support\Arr::toCssClasses([
                             'fi-ta-text-item-icon h-5 w-5',
-                            match ($color) {
+                            match ($iconColor ?? $color) {
```
Comment (zepfietje) [inline packages/tables/resources/views/columns/text-column.blade.php]:
```suggestion
                            match ($iconColor) {
```

File: packages/tables/resources/views/columns/text-column.blade.php [hunk 6]
```diff
@@ -77,23 +77,24 @@
...
+                            match ($iconColor) {
                                 'gray', null => 'text-gray-400 dark:text-gray-500',
                                 default => 'text-custom-500',
                             },
                         ]);
 
                         $iconStyles = \Illuminate\Support\Arr::toCssStyles([
                             \Filament\Support\get_color_css_variables(
-                                $color,
+                                $iconColor ?? $color,
```
Comment (zepfietje) [inline packages/tables/resources/views/columns/text-column.blade.php]:
```suggestion
                                $iconColor,
```

File: packages/tables/resources/views/columns/text-column.blade.php [hunk 7]
```diff
@@ -77,23 +77,24 @@
...
                             },
                         ]);
 
                         $iconStyles = \Illuminate\Support\Arr::toCssStyles([
                             \Filament\Support\get_color_css_variables(
-                                $color,
+                                $iconColor,
                                 shades: [500],
-                            ) => $color !== 'gray',
+                            ) => ($iconColor ?? $color) !== 'gray',
```
Comment (zepfietje) [inline packages/tables/resources/views/columns/text-column.blade.php]:
```suggestion
                            ) => $iconColor !== 'gray',
```

File: packages/tables/docs/03-columns/02-text.md [hunk 8]
```diff
@@ -279,6 +279,20 @@ TextColumn::make('email')
 
 <AutoScreenshot name="tables/columns/text/icon-after" alt="Text column with icon after" version="3.x" />
 
+You may set the color of an icon using `iconColor()`:
```
Comment (zepfietje) [inline packages/tables/docs/03-columns/02-text.md]:
```suggestion
The icon color defaults to the text color, but can be set separately using `iconColor()`:
```

File: packages/tables/docs/03-columns/02-text.md [hunk 9]
```diff
@@ -279,6 +279,20 @@ TextColumn::make('email')
...
 
+The icon color defaults to the text color, but can be set separately using `iconColor()`:
+
+```php
+use Filament\Support\Enums\IconPosition;
+use Filament\Tables\Columns\TextColumn;
+
+TextColumn::make('email')
+    ->icon('heroicon-m-envelope')
+     ->iconPosition(IconPosition::After) // `IconPosition::Before` or `IconPosition::After`
```
Comment (zepfietje) [inline packages/tables/docs/03-columns/02-text.md]:
```suggestion
```

File: packages/tables/docs/03-columns/02-text.md [hunk 10]
```diff
@@ -279,6 +279,20 @@ TextColumn::make('email')
 
 <AutoScreenshot name="tables/columns/text/icon-after" alt="Text column with icon after" version="3.x" />
 
+The icon color defaults to the text color, but can be set separately using `iconColor()`:
+
+```php
+use Filament\Support\Enums\IconPosition;
```
Comment (zepfietje) [inline packages/tables/docs/03-columns/02-text.md]:
```suggestion
```

File: packages/tables/docs/03-columns/02-text.md [hunk 11]
```diff
@@ -279,6 +279,18 @@ TextColumn::make('email')
 
 <AutoScreenshot name="tables/columns/text/icon-after" alt="Text column with icon after" version="3.x" />
 
+The icon color defaults to the text color, but can be set separately using `iconColor()`:
```
Comment (zepfietje) [inline packages/tables/docs/03-columns/02-text.md]:
```suggestion
The icon color defaults to the text color, but you may customize the icon color separately using `iconColor()`:
```

File: packages/tables/docs/03-columns/02-text.md [hunk 12]
```diff
@@ -279,6 +279,18 @@ TextColumn::make('email')
...
+
+```php
+use Filament\Tables\Columns\TextColumn;
+
+TextColumn::make('email')
+    ->icon('heroicon-m-envelope')
+    ->iconColor('success')
+```
+
+<AutoScreenshot name="tables/columns/text/icon-color" alt="Text column with icon color" version="3.x" />
```
Comment (zepfietje) [inline packages/tables/docs/03-columns/02-text.md]:
```suggestion
<AutoScreenshot name="tables/columns/text/icon-color" alt="Text column with icon in the primary color" version="3.x" />
```

File: packages/tables/docs/03-columns/02-text.md [hunk 13]
```diff
@@ -279,6 +279,18 @@ TextColumn::make('email')
...
 <AutoScreenshot name="tables/columns/text/icon-after" alt="Text column with icon after" version="3.x" />
 
+The icon color defaults to the text color, but you may customize the icon color separately using `iconColor()`:
+
+```php
+use Filament\Tables\Columns\TextColumn;
+
+TextColumn::make('email')
+    ->icon('heroicon-m-envelope')
+    ->iconColor('success')
```
Comment (zepfietje) [inline packages/tables/docs/03-columns/02-text.md]:
```suggestion
    ->iconColor('primary')
```

File: packages/infolists/docs/03-entries/02-text.md [hunk 14]
```diff
@@ -246,6 +246,20 @@ TextEntry::make('email')
...
+use Filament\Support\Enums\IconPosition;
+
+TextEntry::make('email')
+    ->icon('heroicon-m-envelope')
+    ->iconPosition(IconPosition::After) // `IconPosition::Before` or `IconPosition::After`
+    ->iconColor('success')
+```
+
+<AutoScreenshot name="infolists/entries/text/icon-color" alt="Text entry with icon color" version="3.x" />
+
```
Comment (zepfietje) [inline packages/infolists/docs/03-entries/02-text.md]:
```suggestion
The icon color defaults to the text color, but you may customize the icon color separately using `iconColor()`:

```php
use Filament\Infolists\Components\TextEntry;

TextEntry::make('email')
    ->icon('heroicon-m-envelope')
    ->iconColor('primary')
```

<AutoScreenshot name="infolists/entries/text/icon-color" alt="Text entry with icon in the primary color" version="3.x" />

```

File: packages/infolists/docs/03-entries/02-text.md [hunk 15]
```diff
@@ -246,6 +246,14 @@ TextEntry::make('email')
...
 <AutoScreenshot name="infolists/entries/text/icon-after" alt="Text entry with icon after" version="3.x" />
 
+The icon color defaults to the text color, but you may customize the icon color separately using `iconColor()`:
+
+```php
+use Filament\Infolists\Components\TextEntry;
+
+TextEntry::make('email')
+    ->icon('heroicon-m-envelope')
+    ->iconColor('primary')
```
Comment (zepfietje) [inline packages/infolists/docs/03-entries/02-text.md:256]:
```suggestion
    ->iconColor('primary')

AutoScreenshot name="infolists/entries/text/icon-color" alt="Text entry with icon in the primary color" version="3.x" />
```

File: packages/infolists/docs/03-entries/02-text.md [hunk 16]
```diff
@@ -246,6 +246,16 @@ TextEntry::make('email')
...
+
+AutoScreenshot name="infolists/entries/text/icon-color" alt="Text entry with icon in the primary color" version="3.x" />
```
Comment (zepfietje) [inline packages/infolists/docs/03-entries/02-text.md]:
```suggestion
<AutoScreenshot name="infolists/entries/text/icon-color" alt="Text entry with icon in the primary color" version="3.x" />
```

File: packages/infolists/docs/03-entries/02-text.md [hunk 17]
```diff
@@ -246,6 +246,16 @@ TextEntry::make('email')
...
+The icon color defaults to the text color, but you may customize the icon color separately using `iconColor()`:
+
+```php
+use Filament\Infolists\Components\TextEntry;
+
+TextEntry::make('email')
+    ->icon('heroicon-m-envelope')
+    ->iconColor('primary')
+
+<AutoScreenshot name="infolists/entries/text/icon-color" alt="Text entry with icon in the primary color" version="3.x" />
```
Comment (zepfietje) [inline packages/infolists/docs/03-entries/02-text.md:258]:
```suggestion
<AutoScreenshot name="infolists/entries/text/icon-color" alt="Text entry with icon in the primary color" version="3.x" />

```

File: packages/infolists/src/Components/Concerns/HasIconColor.php [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        $this->iconColor = $color;
+
+        return $this;
+    }
+
+    /**
+     * @param  array<mixed> | Closure  $colors
+     */
+    public function iconColors(array | Closure $colors): static
```
Comment (zepfietje) [inline packages/infolists/src/Components/Concerns/HasIconColor.php]: I don't think we need this API, do we, @danharrin?

Comment (danharrin) [reply packages/infolists/src/Components/Concerns/HasIconColor.php]: Agree

Comment (aymanalareqi) [reply packages/infolists/src/Components/Concerns/HasIconColor.php]: @zepfietje @danharrin  in my case, I need to make icon color base on record value

File: packages/infolists/src/Components/Concerns/HasIconColor.php [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+            'state' => $state,
+        ]);
+
+        if ($color === false) {
+            return null;
+        }
+
+        if (filled($color)) {
+            return $color;
+        }
```
Comment (zepfietje) [inline packages/infolists/src/Components/Concerns/HasIconColor.php]: Do we really need these checks, @aymanalareqi @danharrin?

Comment (danharrin) [reply packages/infolists/src/Components/Concerns/HasIconColor.php]: No

**Pull request merged**